subsys/bluetooth/controller/hci/hci.c

```diff
@@ -4798,6 +4798,136 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
 	rp->handle = sys_cpu_to_le16(handle);
 }
 #endif /* CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL */
+
+#if defined(CONFIG_BT_HCI_VS_FATAL_ERROR)
+/* A memory pool for vandor specific events for fatal error rerporting purposes. */
+NET_BUF_POOL_FIXED_DEFINE(vs_err_tx_pool, 1, BT_BUF_RX_SIZE, 8, NULL);
+
+/* The alias to for convenience of Controller HCI implementation. Controller is build for
+ * a particular architecture hence the alias will allow to avoid conditional compilation.
+ * Host may be not aware of hardware architecture the Controller is working on, hence
+ * all CPU data types for supported architectures should be available during build.
+ */
+#if defined(CONFIG_CPU_CORTEX_M)
+typedef struct bt_hci_vs_fata_error_cpu_data_cortex_m bt_hci_vs_fatal_error_cpu_data;
+
+static void vs_err_fatal_cpu_data_fill(bt_hci_vs_fatal_error_cpu_data *cpu_data,
+				       const z_arch_esf_t *esf)
+{
+	cpu_data->a1 = sys_cpu_to_le32(esf->basic.a1);
+	cpu_data->a2 = sys_cpu_to_le32(esf->basic.a2);
+	cpu_data->a3 = sys_cpu_to_le32(esf->basic.a3);
+	cpu_data->a4 = sys_cpu_to_le32(esf->basic.a4);
+	cpu_data->ip = sys_cpu_to_le32(esf->basic.ip);
+	cpu_data->lr = sys_cpu_to_le32(esf->basic.lr);
+	cpu_data->xpsr = sys_cpu_to_le32(esf->basic.xpsr);
+}
+#endif /* CONFIG_CPU_CORTEX_M */
+
+static struct net_buf *vs_err_evt_create(uint8_t subevt, uint8_t len)
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&vs_err_tx_pool, K_FOREVER);
+	if (buf) {
+		struct bt_hci_evt_le_meta_event *me;
+		struct bt_hci_evt_hdr *hdr;
+
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = BT_HCI_EVT_VENDOR;
+		hdr->len = len + sizeof(*me);
+
+		me = net_buf_add(buf, sizeof(*me));
+		me->subevent = subevt;
+	}
+
+	return buf;
+}
+
+struct net_buf *hci_vs_err_stack_frame(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* Prepare vendor specific HCI Fatal Error event */
+	struct bt_hci_vs_fatal_error_stack_frame *sf;
+	bt_hci_vs_fatal_error_cpu_data *cpu_data;
+	struct net_buf *buf;
+
+	buf = vs_err_evt_create(BT_HCI_EVT_VS_ERROR_DATA_TYPE_STACK_FRAME,
+				sizeof(*sf) + sizeof(*cpu_data));
+	BT_ASSERT_MSG(buf != NULL, "Fatal error\n");
+
+	sf = net_buf_add(buf, (sizeof(*sf) + sizeof(*cpu_data)));
+	sf->reason = sys_cpu_to_le32(reason);
+	sf->cpu_type = BT_HCI_EVT_VS_ERROR_CPU_TYPE_CORTEX_M;
+
+	vs_err_fatal_cpu_data_fill(
+		(bt_hci_vs_fatal_error_cpu_data *)sf->cpu_data, esf);
+
+	return buf;
+}
+
+static struct net_buf *hci_vs_err_trace_create(uint8_t data_type, const char *file, uint32_t line,
+					       uint64_t pc)
+{
+	uint32_t len = 0U, pos = 0U;
+	struct net_buf *buf = NULL;
+
+	if (file) {
+		while (file[len] != '\0') {
+			if (file[len] == '/') {
+				pos = len + 1;
+			}
```

Comment:
Can you elaborate on why we are updating `pos` on a `/`?

Reply:
Ah, yes that is not clear I'll provide a comment about it.
The `pos` is used because here the file name is extracted from path. 
From `gcc` documentation: 
> __FILE__
This macro expands to the name of the current input file, in the form of a C string constant. **This is the path by which the preprocessor opened the file**, not the short name specified in ‘#include’ or as the input file name argument. For example, "/usr/local/include/myheader.h" is a possible expansion of this macro.

Reply:
Ah, so `file` is the _file path_?

Reply:
yes and it would be better to change a name of the argument to `file_path`.